### ISSUE TITLE
Consistent logging and handling of errors

### ISSUE THREAD

**Issue by matejpavlovic** [filecoin-project/mir#270]
# Proposition

## Logging

Each module that produces any log output should have a logger parameter. Important events during the module's operation should produce a log output. In the context of distributed protocols, the log levels have the following semantics:

### ERROR

This should never happen. An error is logged if and only if the functionality of the system might be compromised and system might not provide the advertised guarantees. This should only be the case if there is a bug in the implementation or if some of the assumptions of the implemented protocol (e.g. on the upper bound of faulty nodes) have been violated.

### WARNING

A warning indicates that something potentially suspicious, but technically legitimate is happening. When a warning is output, the distributed system as a whole is still fully functional, even tough parts of it might not be any more. A warning is the sign of something being sub-optimal, but accounted for. For example, a warning should be triggered by the reception of a malformed message (another node might be faulty), a buffer running out of space and dropping messages (network asynchrony or sub-optimal configuration), a node (local or remote) severely lagging behind (slow node), etc. If all the nodes are correct and operating smoothly (no significant slow-down) and the network does not drop or significantly delay messages, no warnings should be logged.

### INFO

An event that indicates the system making progress in its operation. Info-level log events should be used sparingly, such that the stream of info-level output gives good insight of what the node is currently doing, but is still observable in real time by a human. For example, a block being delivered, the configuration having changed, or a checkpoint being created, all are typical info-level events.

### DEBUG

Debug events provide insight into the internal workings of the system. The debug events can be produced at a higher rate than a human can perceive and are intended for recording and studying later. Interpreting a debug-level log entry can require deep knowledge of the details of the implementation.

### TRACE

Very fine-grained log messages, providing a detailed insight about what code is being executed. Trace-level logging is not expected to occur all throughout the code, but should only be inserted when needed.

## Handling errors

Use the [https://pkg.go.dev/github.com/pkg/errors](https://pkg.go.dev/github.com/pkg/errors) library to create errors throughout the code.
It is old and a third-party one, but the [native Go errors library](https://pkg.go.dev/errors) does not support stack traces.
Such errors can then be conveniently printed (including a stack trace, if present) using the [errstack package](https://github.com/filecoin-project/mir/blob/main/pkg/util/errstack/errstack.go).

**Comment by sergefdrv** [comment]
Have you considered a `Fatal` level of logging with the semantics as in many widely used logging packages, e.g. [log](https://pkg.go.dev/log#Fatal), [zap](https://pkg.go.dev/go.uber.org/zap#Logger.Fatal), [go-logger](https://pkg.go.dev/github.com/apsdehal/go-logger?utm_source=godoc#Logger.Fatal)?

What is the principal difference between `Debug` and `Trace` levels and how to decide which one to use?

**Comment by abread** [comment]
Have you considered standardizing on [tracing](https://opentelemetry.io/docs/concepts/observability-primer/#distributed-traces) rather than logging?

It's a godsend for debugging distributed systems! In the case of Mir, even if we only capture traces from a single node, it could prove very useful to debug interactions where [non-go modules](https://github.com/filecoin-project/mir/discussions/212) are involved

**Comment by matejpavlovic** [comment]
Yes we can have `Fatal` too, good idea. I added it to the description, as well as a differentiation between `Debug` and `Trace`.

@abread , thanks for the pointer. The possibility of tracing is one of the main strengths of Mir (see #220 ). A rudimentary tool is alreaty there (the event interceptor and `mircat`) and is already very useful, but definitely can use an extension.
